### ISSUE TITLE
Unable to use a `Literal` bound on a `TypeVar`

### ISSUE THREAD

**Issue by nh13** [python/mypy#16535]
**Bug Report**

Unable to use a `Literal` bound on a `TypeVar` (related to https://github.com/fulcrumgenomics/fgpyo/issues/73)

**To Reproduce**

```python
from typing import TypeVar

import sys
if sys.version_info >= (3, 8):
    from typing import Literal
else:
    from typing_extensions import Literal

LiteralType = TypeVar("LiteralType", bound="Literal")
```

**Expected Behavior**

No error is reported by mypy (works in 0.971, fails in 1.7.0)

**Actual Behavior**

```shell
 mypy tmp.py 
tmp.py:9: error: Literal[...] must have at least one parameter  [valid-type]
tmp.py:9: error: TypeVar "bound" must be a type  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.7.0
- Mypy command-line flags: NA
- Mypy configuration options from `mypy.ini` (and other config files): NA
- Python version used: 3.9.18

Want to use this type here: https://github.com/fulcrumgenomics/fgpyo/blob/ca4a572524d2e5f79838250603e2fcf4d5d63155/fgpyo/util/types.py#L170

**Comment by erictraut** [comment]
Using `Literal` in this context doesn't make sense. `Literal` is a special form that allows you to specify a type. It's not a type itself, so it cannot be used by itself in a type annotation or a TypeVar upper bound. Mypy is correct in flagging this as an error. 

If your intent is to have a type that represents "a union of all possible string literals", you can use `LiteralString`, which was introduced in [PEP 675](https://peps.python.org/pep-0675/). Mypy doesn't have full support for this PEP yet, but you can see how it works in [pyright](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2ANIcWVQMowipoBQnAJicFMAFcUAYwAUKMCgD6lehUoAuKAGcWtWaXlKimqgG0ARMDBgDAXQCUC9lFtQAHtrlMWbKAF4oE6RoaUoAMRQACqIJFBk4CA2dnBOupTMrOgeUL7ynOxAA) if you're interested in playing with it.

**Comment by nh13** [comment]
I am looking for a type hint that will say “any literal type” , not just a string, since my function takes a callable parameter that itself takes a string and returns an instance of the specific literal form/type. This is analogous to argument parsing (string to specific type) and this code is used in defopt as well (but it doesn’t have type hints). Thanks for your help!

**Comment by erictraut** [comment]
> I am looking for a type hint that will say “any literal type”

The Python type system currently has no way of expressing "the union of all possible literal types". I don't think I've ever seen a use case for such a concept, so this isn't surprising.

The closest thing is `LiteralString`, which means "the union of all possible string literal types". Also, the union of all possible `bool` literals is equivalent to `bool`, and the union of all possible enum literals (for a given enum) is the enum itself.

You can approximate "the union of all possible int literal types" by enumerating the specific integers that you're interested in, as in `OneDigitInts: TypeAlias = Literal[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`.

**Comment by nh13** [comment]
> I don't think I've ever seen a use case for such a concept

I tried adding type hints for the popular [defopt](https://pypi.org/project/defopt/) command line parsing library for this function: https://github.com/anntzer/defopt/blob/dbfb879268c52eefcaa28949e8e2e2e257ebe28f/src/defopt.py#L1146

to my own version here: https://github.com/fulcrumgenomics/fgpyo/blob/ca4a572524d2e5f79838250603e2fcf4d5d63155/fgpyo/util/types.py#L170

the latter is used to inspect the types for attributes in data classes (attr classes) as to rehydrate them from disk (delimited data files). So like passing an argument to the command line when defopt is used, a string value needs to be parsed into an instance. The type hint needed here is to enable parsing of literal attributes.
